src/cloupe_package_nj9/cloupe.py

```diff
@@ -0,0 +1,268 @@
+                celltrack["Metadata"] = self.read_block(
+                    start=celltrack["Metadata"]["Start"],
+                    end=celltrack["Metadata"]["End"],
+                    as_json=True,
+                )
+                vblock_bytes = self.read_block(
+                    start=celltrack["Values"]["Start"],
+                    end=celltrack["Values"]["End"],
+                )
+                vblock_count = celltrack["Values"]["ArraySize"]
+                groups = celltrack["Metadata"]["groups"]
+                group_keys = struct.unpack(f"{vblock_count}h", vblock_bytes)
+                celltrack["Values"] = [groups[gk] if gk!=-1 else "UNASSIGNED" for gk in group_keys]
+                self.celltracks.append(celltrack)
+
+    def read_header(self, position=0):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(position)
+            header = f.read(4096)
+            return json.loads(header.decode("utf-8", errors="replace").strip("\x00"))
+
+    def get_file_size(self):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(0, 2)
+            size = f.tell()
+            return size
+
+    def __repr__(self):
+        return str(self.index_block["Runs"])
+
+    # trick for partial object decompression
+    def decompress(self, chunk):
+        gz = zlib.decompressobj(31)
+        return gz.decompress(chunk)
+
+    # read cloupe block
+    def read_block(self, start, end, as_json=False):
+        with open(self.cloupe_path, "rb") as f:
+            f.seek(start)
+            read_size = end - start
+            block = f.read(read_size)
+            if block[:2] == b"\x1f\x8b":
+                logging.debug("compressed block")
+                block = self.decompress(block)
+            if as_json:
+                return json.loads(block.decode("utf-8", errors="replace"))
+            return block
+
+
+if __name__=="__main__":
+    logging.basicConfig(level=logging.INFO, format="[%(asctime)s][%(levelname)s] %(message)s")
+    # Create the parser
+    parser = argparse.ArgumentParser(description='Cloupe parser which extracts Barcodes, Features and Annotations info')
+    # Add arguments
+    parser.add_argument('input_file', type=str, help='Path to the input file')
+    # Parse the arguments
+    args = parser.parse_args()
+    # Access the arguments
+    # Load cloupe file
+    cwd = os.getcwd()
+    # hardcoding the location of the files that are generated out of the script
+    barcodes = cwd + '/barcodes.csv'
+    features = cwd + '/features.csv'
+    barcode_clusters = cwd + '/annotations.csv'
+    cloupe = Cloupe(args.input_file)
+
+    count_values = {}
+    # Getting the count of Features and Barcodes to set the limits for the columns in the csv files
+    for prop in ['FeatureCount', 'BarcodeCount']:
+        count_values[prop] = cloupe.matrices[0][prop]
+
+
+    collection = []
+    # writing the barcodes to the csv files
+    with open(barcodes, 'w') as barcodes_file:
```

Comment:
Could each one of the writes (barcodes, features, and annotations) be an individual function. If so, should they be inside cloupe?

Have a look at https://docs.python.org/3.9/library/csv.html for writing CSVs with quoting and all that fun.

Reply:
I had been trying various combination of csv writter with the data that we have. I was finding it difficult to dump out the csv properly even when i was using the quoting and quotechar features along with the delimiter. I was thinking if i cud stick without using csv module. Let me know ur thoughts. Regarding moving the writes as methods. My query would be do i remove the main function entirely and the cloupe.py will remain a class with methods to write barcodes, features and annotations is this what u were intending?